### PR TITLE
Dev/duncanmartyn

### PR DESCRIPTION
**Pull request by duncanmartyn**
Added core workshop logic refactored into AML component format, AML workspace setup (data, model, environment) utils, and amended workshop notebook.

### CODE REVIEW

File: .template.env [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# using this file: copy and rename to .env and populate the values for each variable
```
Comment (Slowika) [inline .template.env:1]: Just wanted to confirm: will you create the `.env` file yourself prior to delivering the workshop with credentials you know, or will the participants have to create the `.env` file themselves?

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+[project]
+name = "aurora-introductory-workshop"
+version = "0.1.0"
+description = "Add your description here"
```
Comment (Slowika) [inline pyproject.toml]: Maybe we should put some basic description here. For example: `"Aurora fine-tuning on Azure Machine Learning workshop"`.

File: setup/common/utils.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    ws_name : str
+        Azure ML workspace name
+
+    Raises
+    ------
+    KeyError
+        If any of the required environment variables are not set.
+
+    """
+    sub_id = os.environ["MLFLOW_TRACKING_URI"].split("/")[6]
```
Comment (Slowika) [inline setup/common/utils.py:29]: I think we should mention these environment variables in the README.

File: setup/common/utils.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    ws_name : str
+        Azure ML workspace name
+
+    Raises
+    ------
+    KeyError
+        If any of the required environment variables are not set.
+
+    """
+    sub_id = os.environ["SUBSCRIPTION_ID"]
```
Comment (Slowika) [inline setup/common/utils.py:53]: These are meant to be in the `.env` file. I suggest you mention these ones in the README as well.

File: setup/components/common/utils.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+    cfg : dict[str, Any]
+        Additional keyword arguments to pass to the AuroraPretrained constructor.
+
+    Returns
+    -------
+    model : aurora.AuroraPretrained
+        Loaded Aurora model.
+
+    """
+    strict = bool(cfg.pop("strict", True))
```
Comment (Slowika) [inline setup/components/common/utils.py:65]: Why are we mutating the dictionary here? I appreciate it doesn't make much of a difference, but popping and casting to a `bool` is not necessary. Instead:
```
strict = cfg.get("strict", default=True)
```
Same goes for all of the other instances of using casting to `bool`, e.g. in `use_lora`.
Is it fine if `AuroraPretrained` is given all of `cfg`, including `use_lora` and `strict` entries?

In any case, the cast to `bool` can go away even if you want to pop these items from the dictionary:
```
strict = cfg.pop("strict", True)
```

File: setup/components/training/main.py [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+
+Running in Azure Machine Learning:
+    See setup/components/training/component.py for definition and deployment, and
+    notebooks/0_aurora_workshop.ipynb for example usage.
+
+Key configuration parameters:
+- type: Whether to perform short lead ("short") or autoregressive ("rollout")
+    fine-tuning.
+- mode: Whether to use synthetic data ("test") or real data ("era5").
+- steps: Number of fine-tuning steps to perform.
```
Comment (Slowika) [inline setup/components/training/main.py]: I prefer `epochs` to `steps` in this context, so that nobody confuses it with `rollout_steps`.

File: setup/components/training/main.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+        create_logger,
+        load_model,
+        register_new_variables,
+        validate_common_config,
+    )
+
+LOG = create_logger()
+
+
+def supervised_mae(pred: Batch, target: Batch) -> torch.Tensor:
```
Comment (Slowika) [inline setup/components/training/main.py]: I think it's fine to have it like this for now, but we should have weighted MAE here (as is done in the Aurora paper). The reason for this is that there are 60+ atmospheric variables and only 4 surface variables, so the losses in the atmospheric variables may dominate the losses for the surface variables.
A future PR will address it (I have the changes in my branch).
In the meantime, `supervised_` in the name is superfluous, it can be removed.

File: setup/components/training/main.py [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+                LOG.info(
+                    "Rollout step complete: no=%d, init_time=%s, pred_time=%s",
+                    ro_step,
+                    rollout_batch.metadata.time[0].isoformat(timespec="hours"),
+                    pred.metadata.time[0].isoformat(timespec="hours"),
+                )
+                rollout_batch = update_batch(rollout_batch, pred)
+
+        pred = model.forward(rollout_batch)
+        loss_value = supervised_mae(pred, tgt_batch)
```
Comment (sali1293) [inline setup/components/training/main.py]: Do we need to recreate tgt_batch again. Currently it's using the one created inside for loop

File: setup/components/training/main.py [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+
+    Parameters
+    ----------
+    model : torch.nn.Module
+        Model to freeze.
+
+    """
+    for name, p in model.named_parameters():
+        name_l = name.lower()
+        p.requires_grad = "lora" in name_l
```
Comment (Slowika) [inline setup/components/training/main.py:123]: Do we have any better guarantees on where the substring `lora` appears in the string (so we don't accidentally forget to freeze some other parameters)? Maybe it is always a prefix/suffix?

File: setup/components/training/main.py (hunk 8 above)
Comment (duncanmartyn) [reply setup/components/training/main.py]: I think you're right, good catch. Will amend first thing tomorrow!

File: setup/components/training/main.py [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+    pred : aurora.Batch
+        Model prediction after fine-tuning step.
+    loss_history : list[float]
+        List of loss value floats.
+
+    """
+    loss_history: list[float] = []
+
+    for step in range(steps):
+        init_batch = batch_fn(start_datetime=start_datetime + timedelta(hours=6) * step)
```
Comment (Slowika) [inline setup/components/training/main.py]: I don't think this is the correct choice of batch, and fixing this requires changes in multiple places.

It is fine to select one batch for each epoch of fine-tuning. However, the batches should be chosen randomly over a long time period. We have the following steps in the fine-tuning loop:
- sample a random data point
- run the model to get the model prediction
- normalise predictions and target data
- compute the loss and update model parameters

Consequentially, `init_batch` here should be the ground-truth state that is `6 * x` hours from the `start_datetime`, where `x` is a randomly generated non-negative integer up to some upper bound. This upper bound should be provided as an argument to this component (with a possible default, for example 120).

The `tgt_batch` variable assignment should also be changed; it should be the ground-truth state of the world at time `start_datetime + 6 * (x + 1)`. As it is, the target is always 6h ahead of the `start_datetime`, which is fixed and given as the script argument.

This way, we are constantly fine-tuning with the same target, but we are varying the input. This is wrong.

File: setup/components/training/main.py [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+    Returns
+    -------
+    pred : aurora.Batch
+        Model prediction after fine-tuning step.
+    loss_history : list[float]
+        List of loss value floats.
+
+    """
+    loss_history: list[float] = []
+    init_batch = batch_fn(start_datetime=start_datetime)
```
Comment (sali1293) [inline setup/components/training/main.py]: We update batch within rollout loop but not across rollout steps. So each step re-trains on the same initial window. Not a huge issue for workshop though. Not critical, rising for discussion

File: setup/components/training/main.py [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+
+    for ft_step in range(steps):
+        optimiser.zero_grad(set_to_none=True)
+        rollout_batch = init_batch
+
+        with torch.no_grad():
+            for ro_step in range(rollout_steps - 1):
+                tgt_batch = batch_fn(
```
Comment (Slowika) [inline setup/components/training/main.py]: The way `tgt_batch` is computed and used seems wrong (and it ends up referring to the wrong batch). It is only ever written to inside the `ro_step` loop and it is overwritten in every step. Only its final value is used, and it is used after the `torch.no_grad()` block ends. This means that all but one of these calls does anything.

If `tgt_batch` is only meant to be used outside the `ro_step` loop, then this variable can be assigned to outside of the loop. You can just move the assignment after the `ro_step` loop (but before the call to `supervised_mae`), with `start_datetime` equal to the time in the `rollout_batch` metadata + 6 hours.

File: setup/components/training/main.py (hunk 11 above)
Comment (Slowika) [inline setup/components/training/main.py]: We probably also want to do the random sampling here, just like in the short lead time fine-tuning.
In that case, `init_batch` should be initialised inside the `ft_step` loop, with `start_datetime` chosen randomly like above.

File: setup/components/training/main.py (hunk 7 above)
Comment (sali1293) [inline setup/components/training/main.py]: @Slowika We are not calculating weighted MAE. I suppose it's alright for Workshop but we can just speak on weighted MAE while explaining finetuning

File: setup/notebooks/register_model.ipynb [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "31d0802a",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "c:\\Users\\duncanmartyn\\Documents\\Dev\\aurora-introductory-workshop\\.venv\\Lib\\site-packages\\tqdm\\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
```
Comment (Slowika) [inline setup/notebooks/register_model.ipynb]: You should clear all notebook outputs before merging.

File: setup/components/training/main.py (hunk 11 above)
Comment (duncanmartyn) [reply setup/components/training/main.py]: Yeah, same as the original code in aurora_demo_core.py I'm pretty sure

File: setup/notebooks/load_era5.ipynb [hunk 14]
```diff
@@ -0,0 +1,2200 @@
...
+        "nteract": {
+          "transient": {
+            "deleting": false
+          }
+        }
+      },
+      "source": [
+        "# Loading ERA5\n",
+        "\n",
+        "This notebook serves to subset and load ERA5 data from Google's public ERA5 analysis-ready, cloud-optimised (ARCO) mirror. Here, ERA5 data for the 1940-01-01 to 2025-12-31 period (continually, if irregularly, updated) at hourly frequency is stored in Zarr format. Beyond format, the sole difference between re-gridded data available therein and that through the Copernicus Climate Data Store (CDS) is variable naming: longnames are used in the former and shortnames in the latter.\n",
```
Comment (Slowika) [inline setup/notebooks/load_era5.ipynb]: Instead of "subset and load", maybe "load and filter"?

File: setup/components/training/main.py (hunk 11 above)
Comment (sali1293) [reply setup/components/training/main.py]: > Yeah, same as the original code in aurora_demo_core.py I'm pretty sure

I will fix this

**Review by Slowika — CHANGES_REQUESTED** [review]
Looks mostly good, well done team! A few minor comments and one important point:

Fine-tuning is somewhat incomplete (missing random sampling), and autoregressive rollout fine-tuning has a bug. This should be fixed before the speedrun. The problems and the ways to fix them are described in the comments.

Also, please clear all of the outputs from all of the notebooks.

File: .template.env (hunk 1 above)
Comment (duncanmartyn) [reply .template.env:1]: This is only required for them to recreate the workshop materials (e.g. component deployment) locally, though it's possible to do so remotely (i.e. in an AML compute instance) without any setup required

File: setup/common/utils.py (hunk 3 above)
Comment (duncanmartyn) [reply setup/common/utils.py:29]: They're automatically set in AML compute instances, not user configured

File: setup/common/utils.py (hunk 4 above)
Comment (duncanmartyn) [reply setup/common/utils.py:53]: They're mentioned by proxy in:
.template.env: contains environment variables to be set for local use, particularly relevant to workspace setup. Copy to a new file named `.env` and fill values.

But if you'd rather explicit description then I can add that!

File: setup/components/common/utils.py (hunk 5 above)
Comment (duncanmartyn) [reply setup/components/common/utils.py:65]: It's for typing: .get("strict", default=True / False) doesn't guarantee a boolean return, nor does dict unpacking without popping and casting use_lora guarantee that the value it gets passed is boolean and so both get flagged.

AuroraPretrained cannot be given the strict entry, so it must be popped.

Admittedly it's perhaps more strict on typing than is necessary.

File: setup/components/training/main.py (hunk 9 above)
Comment (duncanmartyn) [reply setup/components/training/main.py:123]: @sali1293 unless you know off the top of your head I can check tomorrow!

File: setup/components/training/main.py (hunk 12 above)
Comment (duncanmartyn) [reply setup/components/training/main.py]: Yes, @sali1293 caught the need to move it outside the loop. When calculating loss outside of the loop in aurora_demo_core it's outside of the no_grad context manager, should that not be the case?

File: setup/components/training/main.py (hunk 9 above)
Comment (duncanmartyn) [reply setup/components/training/main.py:123]: Doesn't look like any named params contain the "lora" substring.

File: setup/components/training/main.py (hunk 10 above)
Comment (duncanmartyn) [reply setup/components/training/main.py]: Understood on the randomly sampled batch, will correct that.

However, the input init_batch is varied as start_datetime + (6h * epoch). The tgt_batch is then the datetime of this batch +6h, so is 6h ahead of the initial state, matching the prediction.

File: setup/components/training/main.py (hunk 9 above)
Comment (duncanmartyn) [reply setup/components/training/main.py:123]: @sali1293 solved, use_lora=True needed to surface named lora params. We could be more specific on the substring used to filter them but doesn't seem necessary

File: setup/components/training/main.py (hunk 7 above)
Comment (duncanmartyn) [reply setup/components/training/main.py]: Resolving with comment: weighted loss will be added in #2
